src/Columns/ColumnVariant.cpp

```diff
@@ -0,0 +1,1370 @@
+}
+
+Field ColumnVariant::operator[](size_t n) const
+{
+    Discriminator discr = localDiscriminatorAt(n);
+    if (discr == NULL_DISCRIMINATOR)
+        return Null();
+    return (*variants[discr])[offsetAt(n)];
+}
+
+void ColumnVariant::get(size_t n, Field & res) const
+{
+    Discriminator discr = localDiscriminatorAt(n);
+    if (discr == NULL_DISCRIMINATOR)
+        res = Null();
+    else
+        variants[discr]->get(offsetAt(n), res);
+}
+
+bool ColumnVariant::isDefaultAt(size_t n) const
+{
+    return localDiscriminatorAt(n) == NULL_DISCRIMINATOR;
+}
+
+bool ColumnVariant::isNullAt(size_t n) const
+{
+    return localDiscriminatorAt(n) == NULL_DISCRIMINATOR;
+}
+
+StringRef ColumnVariant::getDataAt(size_t) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method getDataAt is not supported for {}", getName());
+}
+
+void ColumnVariant::insertData(const char *, size_t)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method insertData is not supported for {}", getName());
+}
+
+void ColumnVariant::insert(const Field & x)
+{
+    if (x.isNull())
+        insertDefault();
+    else
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Cannot insert field {} to column {}", toString(x), getName());
+}
+
+void ColumnVariant::insertFrom(const IColumn & src_, size_t n)
+{
+    const ColumnVariant & src = assert_cast<const ColumnVariant &>(src_);
+
+    const size_t num_variants = variants.size();
+    if (src.variants.size() != num_variants)
+        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert value of Variant type with different number of types");
+
+    /// Remember that src column can have different local variants order.
+    Discriminator global_discr = src.globalDiscriminatorAt(n);
+    Discriminator local_discr = localDiscriminatorByGlobal(global_discr);
+    getLocalDiscriminators().push_back(local_discr);
+    if (local_discr == NULL_DISCRIMINATOR)
+    {
+        getOffsets().emplace_back();
+    }
+    else
+    {
+        getOffsets().push_back(variants[local_discr]->size());
+        variants[local_discr]->insertFrom(src.getVariantByGlobalDiscriminator(global_discr), src.offsetAt(n));
+    }
+}
+
+void ColumnVariant::insertIntoVariant(const DB::Field & x, Discriminator global_discr)
+{
+    if (global_discr > variants.size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid global discriminator: {}. The number of variants is {}", size_t(global_discr), variants.size());
+    auto & variant = getVariantByGlobalDiscriminator(global_discr);
+    variant.insert(x);
+    getLocalDiscriminators().push_back(localDiscriminatorByGlobal(global_discr));
+    getOffsets().push_back(variant.size() - 1);
+}
+
+void ColumnVariant::insertRangeFrom(const IColumn & src_, size_t start, size_t length)
+{
+    const size_t num_variants = variants.size();
+    const auto & src = assert_cast<const ColumnVariant &>(src_);
+    if (src.variants.size() != num_variants)
```

Comment:
AFAIK we should not have variants with different types here. Before each inter-columns insertions we should check that types of columns are the same. We check it by comparing data types or using `structureEquals(...)` method of `IColumn`. This check is added more like to prevent segfaults/undefined behaviour if there is a bug and we actually have different columns. Also if we will have different variant column types, we will anyway get an exception during `insertRangeFrom` on variants.